File: backend/internal/models/permission_report.go
package models

// SPermissionReportRow represents a single row in the Permission Report matrix.
// It joins DBFLPASS (user metadata) with dbmenu or dbmenureport (permissions)
// to produce one human-readable record per user-menu pair.
type SPermissionReportRow struct {
	UserID     string `gorm:"column:USERID"   json:"user_id"`
	Username   string `gorm:"column:USERID"   json:"username"`
	FullName   string `gorm:"column:FullName" json:"full_name"`
	Role       string `gorm:"column:role"     json:"role"`
	MenuCode   string `gorm:"column:KODEMENU" json:"menu_code"`
	MenuName   string `gorm:"column:Keterangan" json:"menu_name"`
	MenuType   string `gorm:"column:menu_type"  json:"menu_type"` // "menu" or "report"
	HasAccess  int    `gorm:"column:HASACCESS"  json:"has_access"`
	IsCreate   int    `gorm:"column:ISTAMBAH"   json:"is_create"`
	IsUpdate   int    `gorm:"column:ISKOREKSI"  json:"is_update"`
	IsDelete   int    `gorm:"column:ISHAPUS"    json:"is_delete"`
	IsPrint    int    `gorm:"column:ISCETAK"    json:"is_print"`
	IsExport   int    `gorm:"column:ISEXPORT"   json:"is_export"`
	IsApprove1 int    `gorm:"column:IsOtorisasi1" json:"is_approve_1"`
	IsApprove2 int    `gorm:"column:IsOtorisasi2" json:"is_approve_2"`
	IsApprove3 int    `gorm:"column:IsOtorisasi3" json:"is_approve_3"`
	IsApprove4 int    `gorm:"column:IsOtorisasi4" json:"is_approve_4"`
	IsApprove5 int    `gorm:"column:IsOtorisasi5" json:"is_approve_5"`
	IsCanceled int    `gorm:"column:IsBatal"    json:"is_canceled"`
}

File: src/domains/reports/permissionReport.ts
/**
 * Permission Report types.
 *
 * These mirror the backend DTO at:
 *   backend/internal/dto/report_dto.go - SPermissionReportRowDTO / SPermissionReportResponse / SReportFilters
 *
 * Field names use snake_case to match the backend JSON tags.
 */

/**
 * A single row in the permission report matrix.
 * Each row represents one (user, menu) pair plus all granular permission flags.
 */
export interface IPermissionReportRow {
  user_id: string
  username: string
  full_name: string
  role: string
  menu_code: string
  menu_name: string
  menu_type: string // "menu" or "report"
  has_access: number
  is_create: number
  is_update: number
  is_delete: number
  is_print: number
  is_export: number
  is_approve_1: number
  is_approve_2: number
  is_approve_3: number
  is_approve_4: number
  is_approve_5: number
  is_canceled: number
}

/**
 * Paginated JSON response for /admin/reports/permissions?format=json
 */
export interface IPermissionReportResponse {
  data: IPermissionReportRow[]
  total: number
  page: number
  per_page: number
}

/**
 * Menu type filter — controls which underlying table is queried on the backend.
 *  - "menu"   → dbmenu (default)
 *  - "report" → dbmenureport
 *  - "all"    → both (UI-only convenience; the BFF treats empty/missing as "all")
 */
export type TMenuTypeFilter = 'menu' | 'report' | 'all'

/**
 * The filter shape consumed by the permission report service. All fields are
 * optional from the UI's perspective; only `menu_type` has a sensible default.
 */
export interface IPermissionReportFilters {
  user_id?: string
  menu_id?: string
  menu_type?: TMenuTypeFilter
  page?: number
  per_page?: number
}
